File: 2020-10-10 - Episodio 5: Más problemas con matrices/0xa6 - Submatriz de máxima suma/reference/n3.py
def kadane(array):
  best = array[0]
  dp = [0] * len(array)
  for i, _ in enumerate(array):
    dp[i] = array[i]
    if i-1 >= 0:
      dp[i] = max(dp[i], dp[i-1] + array[i])
    best = max(best, dp[i])
  return best

rows, cols = [int(x) for x in input().split()]
a = []
for _ in range(rows):
  a.append([int(x) for x in input().split()])

answer = a[0][0]
for r0 in range(rows):
  compressed = [0] * cols
  for r1 in range(r0, rows):
    for c in range(cols):
      compressed[c] += a[r1][c]
    answer = max(answer, kadane(compressed))

print(answer)

File: 2020-10-10 - Episodio 5: Más problemas con matrices/0xa6 - Submatriz de máxima suma/reference/n3.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <algorithm>
using namespace std;

const int MAXN = 50;

int a[MAXN][MAXN];

int main() {
  int rows, cols;
  cin >> rows >> cols;
  assert(1 <= rows and rows <= 50);
  assert(1 <= cols and cols <= 50);

  for (int i = 0; i < rows; ++i) {
    for (int j = 0; j < cols; ++j) {
      cin >> a[i][j];
    }
  }

  int answer = a[0][0];

  for (int r0 = 0; r0 < rows; ++r0) {
    vector<int> sum(cols, 0);
    for (int r1 = r0; r1 < rows; ++r1) {
      for (int c = 0; c < cols; ++c) {
        sum[c] += a[r1][c];
      }
      // Now do Kadane's algorithm om sum
      // dp[i] = max(dp[i-1] + a[i], a[i])
      vector<int> dp(cols);
      int best = sum[0];
      for (int c = 0; c < cols; ++c) {
        dp[c] = sum[c];
        if (c-1 >= 0) dp[c] = max(dp[c], dp[c-1] + sum[c]);
        best = max(best, dp[c]);
      }
      answer = max(answer, best);
    }
  }

  cout << answer << endl;
  return 0;
}